### PR TITLE
Matlab cmd as script

### PR DESCRIPTION
**Pull request by nbhoski**
* Added support to run MATLAB command through script file to address the issues of "" and other special characters.
* Writing the command in command_UUID.m script file and passing that as a parameter to run_matalb_command(bat|sh) script .
* cleaning the entire temp folder after job completion.
* Few tests which were based on console output of the command has been disabled as now commands cannot be seen on the console window as it will be wrapped in temp script file . I shall add them back to integ-tests once integ-test folders merged.

### CODE REVIEW

File: src/main/java/com/mathworks/ci/MatlabBuild.java [hunk 1]
```diff
@@ -22,33 +22,29 @@
...
             throws IOException, InterruptedException {
         // Get node specific tmp directory to copy matlab runner script
         String tmpDir = getNodeSpecificTmpFolderPath(workspace);
         FilePath targetWorkspace = new FilePath(launcher.getChannel(), tmpDir);
-        ProcStarter matlabLauncher;
         if (launcher.isUnix()) {
             final String runnerScriptName = uniqueName + "/run_matlab_command.sh";
-            matlabLauncher = launcher.launch().pwd(workspace).envs(envVars)
-                    .cmds(tmpDir + "/" + runnerScriptName, matlabCommand).stdout(listener);
+            matlabLauncher.cmds(tmpDir + "/" + runnerScriptName, matlabCommand).stdout(listener);
```
Comment (mcafaro) [inline src/main/java/com/mathworks/ci/MatlabBuild.java:40]: Why does the `runnerScriptName` start with a forward slash and then `tmpDir` is separated from it with a forward slash as well?

Comment (nbhoski) [reply src/main/java/com/mathworks/ci/MatlabBuild.java:40]: That's because I create a tmpDir as UUID in keep it in system tmp directory with all helper files like runner script and now Matlab script and delete the entire folder at the end.

File: src/main/java/com/mathworks/ci/RunMatlabCommandBuilder.java [hunk 2]
```diff
@@ -112,22 +116,40 @@ public void perform(@Nonnull Run<?, ?> build, @Nonnull FilePath workspace,
     private synchronized int execMatlabCommand(FilePath workspace, Launcher launcher,
             TaskListener listener, EnvVars envVars) throws IOException, InterruptedException {
         final String uniqueTmpFldrName = getUniqueNameForRunnerFile();
-        ProcStarter matlabLauncher;
+        final String uniqueCommandFile =
+                "command_" + getUniqueNameForRunnerFile().replaceAll("-", "_");
+
+        // Get unique temporary folder filepath
+        final FilePath uniqeTmpFolderPath =
```
Comment (mw-akumar) [inline src/main/java/com/mathworks/ci/RunMatlabCommandBuilder.java:123]: At the very least the creation of the temporary command file should be its own function so that we don't have all these implementation details here.

File: src/main/java/com/mathworks/ci/RunMatlabCommandBuilder.java [hunk 3]
```diff
@@ -112,22 +116,40 @@ public void perform(@Nonnull Run<?, ?> build, @Nonnull FilePath workspace,
...
+        final FilePath matlabCommandFile =
+                new FilePath(uniqeTmpFolderPath, uniqueCommandFile + ".m");
+        final String matlabCommandFileContent =
+                "cd '" + workspace.getRemote().replaceAll("'", "''") + "';\n" + getCommand();
+
+        // Display the commands on console output for users reference
+        listener.getLogger()
+                .println("Generating MATLAB script with content:\n" + getCommand() + "\n");
+
+        matlabCommandFile.write(matlabCommandFileContent, "UTF-8");
```
Comment (mw-akumar) [inline src/main/java/com/mathworks/ci/RunMatlabCommandBuilder.java:135]: MATLAB supports UTF-8 since 20a. Versions prior to 20a may not work correctly if we have non-ASCII characters in the command. We may have to document limitations regarding non-ASCII characters.

File: src/main/java/com/mathworks/ci/RunMatlabTestsBuilder.java [hunk 4]
```diff
@@ -238,14 +238,18 @@ private synchronized int execMatlabCommand(FilePath workspace, Launcher launcher
         final String uniqueTmpFldrName = getUniqueNameForRunnerFile();
         ProcStarter matlabLauncher;
         try {
-            matlabLauncher = getProcessToRunMatlabCommand(workspace, launcher, listener, envVars,
-                    constructCommandForTest(getInputArguments()), uniqueTmpFldrName);
+            // Get decorated launcher for Windows
+            launcher = getDecoratedLauncherForWindows(launcher);
```
Comment (mw-akumar) [inline src/main/java/com/mathworks/ci/RunMatlabTestsBuilder.java:242]: The decoration of cmd for Windows was earlier encapsulated by the code that created the ProcStarter. Now it has leaked into the execMATLABCommand. Why is this needed?

File: src/main/java/com/mathworks/ci/RunMatlabCommandBuilder.java (hunk 2 above)
Comment (nbhoski) [reply src/main/java/com/mathworks/ci/RunMatlabCommandBuilder.java:121]: Done. file creation is moved to separate method.

File: src/main/java/com/mathworks/ci/RunMatlabTestsBuilder.java (hunk 4 above)
Comment (nbhoski) [reply src/main/java/com/mathworks/ci/RunMatlabTestsBuilder.java]: Removed decoration usage as it was coming from Launcher now handling it in cmds() method now .

File: src/main/java/com/mathworks/ci/MatlabBuild.java [hunk 5]
```diff
@@ -22,33 +22,29 @@
...
      * @param workspace Current build workspace
      * @param launcher Current build launcher
      * @param listener Current build listener
-     * @param envVars Environment variables of the current build
      * @param matlabCommand MATLAB command to execute on shell
      * @return matlabLauncher returns the process launcher to run MATLAB commands
      */
-    default ProcStarter getProcessToRunMatlabCommand(FilePath workspace, Launcher launcher,
-            TaskListener listener, EnvVars envVars, String matlabCommand, String uniqueName)
+    default ProcStarter getProcessToRunMatlabCommand(ProcStarter matlabLauncher, FilePath workspace,
```
Comment (mw-akumar) [inline src/main/java/com/mathworks/ci/MatlabBuild.java:32]: This interface change for getProcessToRunMatlabCommand still seems a bit odd and unnecessary. The callers are creating the ProcStarter object, making a change to the cwd, and then calling getProcessToRunMatlabCommand. Instead can we
1. pass the cwd to getProcessToRunMatlabCommand as a parameter. This way the construction of the ProcStarter method is still self contained in one method. 
or
2. set the cwd on the ProcStarter object after calling getProcessToRunMatlabCommand. This way, there is no need to change the interface.

File: src/main/java/com/mathworks/ci/RunMatlabCommandBuilder.java [hunk 6]
```diff
@@ -112,22 +116,45 @@ public void perform(@Nonnull Run<?, ?> build, @Nonnull FilePath workspace,
...
+        final FilePath matlabCommandFile =
+                new FilePath(uniqeTmpFolderPath, uniqueScriptName + ".m");
+        final String matlabCommandFileContent =
+                "cd '" + workspace.getRemote().replaceAll("'", "''") + "';\n" + getCommand();
+
+        // Display the commands on console output for users reference
+        listener.getLogger()
+                .println("Generating MATLAB script with content:\n" + getCommand() + "\n");
+
+        matlabCommandFile.write(matlabCommandFileContent, "UTF-8");
```
Comment (mw-akumar) [inline src/main/java/com/mathworks/ci/RunMatlabCommandBuilder.java:158]: There is still the open question of UTF-8 support for MATLAB in older releases. I am okay if we just document the limitation.

Comment (nbhoski) [reply src/main/java/com/mathworks/ci/RunMatlabCommandBuilder.java:158]: I will discuss with @mw-hrastega and make sure if its get documented at the least

File: src/main/java/com/mathworks/ci/MatlabBuild.java (hunk 5 above)
Comment (nbhoski) [reply src/main/java/com/mathworks/ci/MatlabBuild.java:32]: Implemented option 2  its much feasible than option 1 . & yes i could avoid many changes with this.

File: src/main/java/com/mathworks/ci/RunMatlabCommandBuilder.java (hunk 6 above)
Comment (acampbel) [reply src/main/java/com/mathworks/ci/RunMatlabCommandBuilder.java:156]: Will a UTF-8 encoded file with ascii characters still work in older releases? Or are we just talking abut the case where someone includes a non-ascii character in the command? If its the latter then I  am fine with the limitation. If its the former that is a blocking issue. If we are writing ascii characters to the file as UTF-8 will it work in older releases? If not we should encode it differently

Comment (mcafaro) [reply src/main/java/com/mathworks/ci/RunMatlabCommandBuilder.java:156]: My understanding is Unicode if fully backward-compatible with ASCII (ASCII is a subset of Unicode). So UTF8 files should work fine in older releases as long as the script contains only ASCII characters.

If a user enters a command with a non-ASCII character, it will not work correctly in older releases. But it will not work correctly in older releases even if we write the file as ASCII.

Comment (acampbel) [reply src/main/java/com/mathworks/ci/RunMatlabCommandBuilder.java:156]: I am good writing it as unicode as long as it is qualified against at least one release where MATLAB didn't support it for m-files.

Comment (acampbel) [reply src/main/java/com/mathworks/ci/RunMatlabCommandBuilder.java:156]: it will be a great feature for new releases to have a Japanese hello world for example

Comment (mw-akumar) [reply src/main/java/com/mathworks/ci/RunMatlabCommandBuilder.java:156]: I demonstrated the implications to Nikhil today. First of all, if the users are using ASCII text only then all this is moot. All our supported encodings use the same binary representation for ASCII. 

The problem occurs if the users use non-ASCII characters. For instance a MATLAB user in Japan who uses Japanese characters in their command. e.g. 'disp("こんにちは"). If the user in Japan is using a Japanese locale Windows machine (a very likely case), then MATLAB prior to R2020a will expect the .m files to be encoded the SHIFT-JIS encoding. Executing the UTF-8 encoded file will produce character corruption. In MATLAB 2020a and newer, the utf-8 encoded file with Japanese characters will be executed correctly.
However, this is not a trivial problem to solve for 19b and older, as we will have to ask MATLAB what the right .m encoding (Shift-JIS on a Japanese Windows,  ISO-8859-1 on an English Windows and so on) will be. For this we will have to actually run MATLAB and execute some commands. :-)

Comment (nbhoski) [reply src/main/java/com/mathworks/ci/RunMatlabCommandBuilder.java:156]: Updated the help doc for custom MATLAB command 

image

File: src/main/resources/com/mathworks/ci/RunMatlabCommandBuilder/help-matlabCommand.html [hunk 7]
```diff
@@ -4,7 +4,10 @@
  <b>Run commands:</b> results = runtests(‘IncludingSubfolders’,true); assert(all(~[results.Failed]))<br>
  <b>Run a script:</b> runMyScript<br>
  <br>&nbsp;</br>
- <b>Recommendation:</b>If you require a number of MATLAB commands to execute your build, consider writing a MATLAB script and executing the script file instead.<br>
+ <b>Recommendation:</b><ul><li>If you require a number of MATLAB commands to execute your build, consider writing a MATLAB script and executing the script file instead.<br>
+ </li>
+ <li>Input text box supports commands only in ASCII formt. For locale specific commands, Write a MATLAB script using the supported chracater encoding and execute the script file.</li>
+ </ul>
  <b>Note:</b> The build will fail if the execution of any MATLAB command causes an error.
```
Comment (nbhoski) [inline src/main/resources/com/mathworks/ci/RunMatlabCommandBuilder/help-matlabCommand.html:11]: @mw-hrastega request you to review these help text changes for locale-specific MATLAB command

Comment (mw-hrastega) [reply src/main/resources/com/mathworks/ci/RunMatlabCommandBuilder/help-matlabCommand.html]: How about something like this?

The **Command** box supports only ASCII characters. If you use non-ASCII characters in your commands, write your commands to a .m file and invoke the script instead.

Comment (mw-hrastega) [reply src/main/resources/com/mathworks/ci/RunMatlabCommandBuilder/help-matlabCommand.html]: Can we remove the existing recommendation? I find it kind of redundant now.

Comment (nbhoski) [reply src/main/resources/com/mathworks/ci/RunMatlabCommandBuilder/help-matlabCommand.html]: Done.

image

Comment (mw-hrastega) [reply src/main/resources/com/mathworks/ci/RunMatlabCommandBuilder/help-matlabCommand.html]: Quick question. Can we also specify a .mlx file? If so, let's change ".m file" to "plain text file".

Comment (nbhoski) [reply src/main/resources/com/mathworks/ci/RunMatlabCommandBuilder/help-matlabCommand.html]: I think we could run both , However instead of plain text we could use .m/.mlx ?

Comment (mw-hrastega) [reply src/main/resources/com/mathworks/ci/RunMatlabCommandBuilder/help-matlabCommand.html]: Yes, that would be better. Lets replace ".m" with ".m or .mlx".

Comment (nbhoski) [reply src/main/resources/com/mathworks/ci/RunMatlabCommandBuilder/help-matlabCommand.html]: changed it to .m or .mlx 


image
